app/src/main/java/org/mozilla/fenix/HomeActivity.kt

```diff
@@ -134,7 +130,10 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
         WebExtensionPopupFeature(components.core.store, ::openPopup)
     }
 
+    private lateinit var navGraphJob : Deferred<Unit>
+
     private val navHost by lazy {
+        loadNavFragment()
```

Comment:
I'm concerned about some aspects of the current implementation:
- we're not giving `navHost` the maximum amount of time to inflate before we use it. Currently, we're blocking 35ms in a local cold VIEW profile I took https://share.firefox.dev/39QZpuH We should reorder the methods inside `onCreate` (i.e. navHost methods come last) to optimize. Also, can we start this process earlier? e.g. create a  NavHostFragment async in `FenixApplication` and attach the fragment in `HomeActivity` when it's ready. I wonder if that would prevent us from blocking at all?
- it's easy to introduce new regressions or crashes because 1) it's unclear that `navHost` will block if it's not ready yet, 2) you can access `navHost` via the view instead (`container`), and 3) this functionality isn't well encapsulated

I'm thinking we do something like:
```kotlin
/** encapsulates async creation of NavHostFragment. Comment explaining why we did it */
class NavHostFragmentAsyncProvider

fun FenixApplication.onCreate {
    startEarlyAsyncStuff()
}

fun FenixApplication.startEarlyAsyncStuff() {
    startAsyncMegazord()
    startAsyncNavHostFragmentProvider()
}

// HomeActivity.kt...
private lateinit var navHost: NavHostFragment

fun HomeActivity.onCreate() {
    // code not requiring navHost
    navHost = components.navHostFragmentAsyncProvider.getBlocking()
    fragmentManager.beginTransaction().add(... // hopefully not underperformant
    // code requiring navHost
}
```

I like that it encapsulates the async code, features a consistent barrier point (making it easier to follow in profiles, harder to mis-use, and easier to refactor), and maximizes the time navHost. It'll also make for more performant WARM starts since we don't need to reinflate the nav graph.

@MarcLeclair What do you think? I'm concerned that my suggestion is "too perfect" but then again async code is really hard to maintain so... 🤷

Reply:
I like the suggestion. I'm wondering though, what did you mean by reordering the method?  I see that I tried to give it as much time to complete (I think I put it right before `onCreate` ends). I do see what you're pointing out on the profile so I'll try to change it up and add the encapsulation too.